File: client/src/plugins/LobbyHub.js
import { HubConnectionBuilder, HttpTransportType, LogLevel } from '@aspnet/signalr'

export default {
  install (Vue) {
    const connection = new HubConnectionBuilder()
      .withUrl('https://localhost:5001/lobbyhub', {
        skipNegotiation:true,
        transport: HttpTransportType.WebSockets
      })
      .configureLogging(LogLevel.Information)
      .build() 

      const lobbyHub = new Vue() 

      Vue.prototype.$lobbyHub = lobbyHub

      
    
      lobbyHub.joinLobby = (id) =>{
        return startedPromise
        .then(() => connection.invoke('JoinLobby', id))
        .catch(console.error)
      }
      
      lobbyHub.leaveLobby = (id) =>{
        return startedPromise
        .then(() => connection.invoke('LeaveLobby', id))
        .catch(console.error)
      }

      lobbyHub.lobbyChange = () =>{
        return startedPromise
        .then(() => connection.invoke('LobbyChange'))
        .catch(console.error)
      }
      
     connection.on('LobbyChange', () => {
        lobbyHub.$emit('lobby-change')
     })
    

     lobbyHub.newPlayer = (lobbyId, playerId) =>{
      return startedPromise
      .then(() => connection.invoke('NewPlayer', lobbyId, playerId))
      .catch(console.error)
      }

      connection.on('NewPlayer', (lobbyId, playerId) => {
        lobbyHub.$emit('new-player', lobbyId, playerId)
      })

      lobbyHub.hostChange = (lobbyId) =>{
        return startedPromise
        .then(() => connection.invoke('HostChange', lobbyId))
        .catch(console.error)
        }
  
        connection.on('HostChange', (lobbyId) => {
          lobbyHub.$emit('host-change', lobbyId)
        })

        lobbyHub.newGame = (lobbyId) =>{
          return startedPromise
          .then(() => connection.invoke('NewGame', lobbyId))
          .catch(console.error)
          }
    
        connection.on('NewGame', (lobbyId) => {
          lobbyHub.$emit('new-game', lobbyId)
        })

        lobbyHub.opponentLeft = (lobbyId) =>{
          return startedPromise
          .then(() => connection.invoke('OpponentLeft', lobbyId))
          .catch(console.error)
          }
    
        connection.on('OpponentLeft', (lobbyId) => {
          lobbyHub.$emit('opponent-left', lobbyId)
        })

        lobbyHub.gameAccepted = (lobbyId) =>{
          return startedPromise
          .then(() => connection.invoke('GameAccepted', lobbyId))
          .catch(console.error)
          }
    
        connection.on('GameAccepted', (lobbyId) => {
          lobbyHub.$emit('game-accepted', lobbyId)
        })

      let startedPromise = null
      function start () {
        startedPromise = connection.start().catch(err => {
          console.error('Failed to connect with hub', err)
          return new Promise((resolve, reject) => 
          setTimeout(() => start().then(resolve).catch(reject), 20000))
        })
        return startedPromise
      }
      connection.onclose(() => start())
      
      start()
  }
}

File: server/Chickencoop.App/Hubs/LobbyHub.cs
using Chickencoop.App.Hubs.Interfaces;
using Chickencoop.Models.Dtos.UpdateDtos;
using Microsoft.AspNetCore.SignalR;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Chickencoop.App.Hubs
{
    public class LobbyHub : Hub<ILobbyHub>
    {
        public async Task JoinLobby(Guid lobbyId) //adds user to the group
        {
            await Groups.AddToGroupAsync(Context.ConnectionId, lobbyId.ToString());
        }

        public async Task LeaveLobby(Guid lobbyId) //removes user from the group
        {
            await Groups.RemoveFromGroupAsync(Context.ConnectionId, lobbyId.ToString());
        }

        public async Task NewPlayer(Guid lobbyId, Guid playerId) //sends message that a new player has joined the lobby
        {
            await Clients.Group(lobbyId.ToString()).NewPlayer(lobbyId, playerId);
            await LobbyChange();
        }

        public async Task LobbyChange() //sends message that a new game was created or that a new player joined or left a lobby
        {
            await Clients.All.LobbyChange();
        }

        public async Task HostChange(Guid lobbyId) //sends message to other player in the game that the host has left
        {
            await Clients.OthersInGroup(lobbyId.ToString()).HostChange(lobbyId);
            await LobbyChange();
        }

        public async Task NewGame(Guid lobbyId) //message to start a new game
        {
            await Clients.OthersInGroup(lobbyId.ToString()).NewGame(lobbyId);
            //yawait LobbyChange();
        }

        public async Task OpponentLeft(Guid lobbyId) //sends a message that opponent left the lobby
        {
            await Clients.OthersInGroup(lobbyId.ToString()).OpponentLeft(lobbyId);
            await LobbyChange();
        }
        
        public async Task GameAccepted(Guid lobbyId) //sends a signal that the new game request was accepted
        {
            await Clients.Group(lobbyId.ToString()).GameAccepted(lobbyId);
        }
    }
}
